synapse/handlers/device.py

```diff
@@ -367,6 +380,20 @@ async def check_device_registered(
 
         raise errors.StoreError(500, "Couldn't generate a device ID.")
 
+    async def _delete_stale_devices(self) -> None:
+        """Background task that deletes devices which haven't been accessed for more than
+        a configured time period.
+        """
+        # We should only be running this job if the config option is defined.
+        assert self._delete_stale_devices_after is not None
+        now_ms = self.clock.time_msec()
+        devices = await self.store.get_devices_not_accessed_since(
+            now_ms - self._delete_stale_devices_after
+        )
+
+        for device in devices:
+            await self.delete_device(device["user_id"], device["device_id"])
```

Comment:
Arg, it has always confused me how we have both a `delete_device` and `delete_devices`... not a huge deal but would it be more efficient to have `get_devices_not_accessed_since` return a map of user ID to list of device IDs, and then call it once per user:

```python
for user_id, device_ids:
    await self.delete_devices(user_id, device_ids)
```

Reply:
Oh good point, I always forget `delete_devices` is a thing (and yeah, boo for almost identical function names).